protocols/autonatv2/src/client/handler/dial_request.rs

```diff
@@ -0,0 +1,342 @@
+use futures::{channel::oneshot, AsyncRead, AsyncWrite};
+use futures_bounded::FuturesSet;
+use libp2p_core::{
+    upgrade::{DeniedUpgrade, ReadyUpgrade},
+    Multiaddr,
+};
+
+use libp2p_swarm::{
+    handler::{
+        ConnectionEvent, DialUpgradeError, FullyNegotiatedOutbound, OutboundUpgradeSend,
+        ProtocolsChange,
+    },
+    ConnectionHandler, ConnectionHandlerEvent, Stream, StreamProtocol, StreamUpgradeError,
+    SubstreamProtocol,
+};
+use scc::hash_cache::DEFAULT_MAXIMUM_CAPACITY;
+use std::{
+    collections::VecDeque,
+    convert::identity,
+    io,
+    iter::{once, repeat},
+    task::{Context, Poll},
+};
+
+use crate::{
+    generated::structs::{mod_DialResponse::ResponseStatus, DialStatus},
+    request_response::{
+        DialDataRequest, DialDataResponse, DialRequest, DialResponse, Request, Response,
+        DATA_FIELD_LEN_UPPER_BOUND, DATA_LEN_LOWER_BOUND, DATA_LEN_UPPER_BOUND,
+    },
+    REQUEST_PROTOCOL_NAME, REQUEST_UPGRADE,
+};
+
+use super::DEFAULT_TIMEOUT;
+
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum Error {
+    #[error("io error")]
+    Io(#[from] io::Error),
+    #[error("invalid referenced address index: {index} (max number of addr: {max})")]
+    InvalidReferencedAddress { index: usize, max: usize },
+    #[error("data request too large: {len} (max: {max})")]
+    DataRequestTooLarge { len: usize, max: usize },
+    #[error("data request too small: {len} (min: {min})")]
+    DataRequestTooSmall { len: usize, min: usize },
+    #[error("timeout")]
+    Timeout(#[from] futures_bounded::Timeout),
+    #[error("server rejected dial request")]
+    ServerRejectedDialRequest,
+    #[error("server chose not to dial any provided address")]
+    ServerChoseNotToDialAnyAddress,
+    #[error("server ran into an internal error")]
+    InternalServer,
+    #[error("server did not respond correctly to dial request")]
+    InvalidResponse,
+    #[error("server was unable to connect to address: {addr:?}")]
+    UnableToConnectOnSelectedAddress { addr: Option<Multiaddr> },
+    #[error("server experienced failure during dial back on address: {addr:?}")]
+    FailureDuringDialBack { addr: Option<Multiaddr> },
+    #[error("error during substream upgrad")]
+    SubstreamError(
+        #[from] StreamUpgradeError<<ReadyUpgrade<StreamProtocol> as OutboundUpgradeSend>::Error>,
+    ),
```

Comment:
Most of the variants of `StreamUpgradeError` can also be mapped to an `io::Error` (apart from `NegotiationFailed` perhaps which should be unreachable in practise in our case because we don't talk to servers that haven't previously advertised their autonat protocol). A non-conforming server may still reject the protocol though so we can't actually use `unreachable!`. But it should be fine to map it to an `io::Error` or something else that expresses "unexpected" error.

Reply:
But it's so convenient.... but I will change it